### ISSUE TITLE
assert_type() not matching function type

### ISSUE THREAD

**Issue by DiegoBaldassarMilleuno** [python/mypy#15153]
**Bug Report**

'assert_type()' refuses to match a function to its own type.

**To Reproduce**

```python
from typing import assert_type, Callable
def myfunc(arg: int) -> None: pass
assert_type(myfunc, Callable[[int], None])

# Gist URL: https://gist.github.com/mypy-play/50dde8e891d971d823c4c6af2e639834
# Playground URL: https://mypy-play.net/?mypy=latest&python=3.11&gist=50dde8e891d971d823c4c6af2e639834
```

**Expected Behavior**

No error.

**Actual Behavior**

```
main.py:5: error: Expression is of type "Callable[[int], None]", not "Callable[[int], None]"  [assert-type]
Found 1 error in 1 file (checked 1 source file)
```

The error disappears if the function takes no arguments.

**Your Environment**

- Mypy version used: both 1.2.0 and current master branch (commit [29cc561](https://github.com/python/mypy/commit/29cc5614b94d2357794fe5c62c64525fd8087e59))
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): default
- Python version used: 3.11

**Comment by erictraut** [comment]
The two types are not equivalent. The function `myfunc` accepts a keyword argument for a parameter named `arg`, but `Callable[[int], None]` accepts an argument by position only. If you add a positional-only parameter marker to `myfunc`, they will be equivalent.

```python
def myfunc(arg: int, /) -> None:
    pass
```

**Comment by ikonst** [comment]
What should we show in the error message, given there's no special form for this?

**Comment by AlexWaygood** [comment]
@erictraut is correct, but the error message mypy gives here isn't particularly helpful. Let's keep this issue open to track improving the error message.

**Comment by erictraut** [comment]
FWIW, here's the error message that pyright emits:
```
"assert_type" mismatch: expected "(int) -> None" but received "(arg: int) -> None"
```
In pyright's error messages, I output a callable syntax that is based on the proposed (but unfortunately rejected) [PEP 677](https://peps.python.org/pep-0677/). Not sure if this is something mypy maintainers would consider adopting.

**Comment by JelleZijlstra** [comment]
It's definitely an error message bug, since we mention the same string on both sides.

**Comment by madt2709** [comment]
This has been fixed on the master branch. The error message now shows: 

```
main.py:5: error: Expression is of type "Callable[[Arg(int, 'arg')], None]", not "Callable[[int], None]"  [assert-type]
Found 1 error in 1 file (checked 1 source file)
```

You can see this in the playground: https://mypy-play.net/?mypy=master&python=3.11&gist=50dde8e891d971d823c4c6af2e639834

**Comment by ikonst** [comment]
Apparently fixed by #15184.

@AlexWaygood, can we close?
